watertap/unit_models/tests/unit_test_harness.py

```diff
@@ -0,0 +1,183 @@
+            raise UnitValueError(
+                "The number of constraints were {num}, but {num_test} was "
+                "expected ".format(
+                    num=number_total_constraints(blk),
+                    num_test=stats["number_total_constraints"],
+                )
+            )
+        if number_unused_variables(blk) != stats["number_unused_variables"]:
+            raise UnitValueError(
+                "The number of unused variables were {num}, but {num_test} was "
+                "expected ".format(
+                    num=number_unused_variables(blk),
+                    num_test=stats["number_unused_variables"],
+                )
+            )
+
+    @pytest.mark.unit
+    def test_units_consistent(self, frame_unit):
+        assert_units_consistent(frame_unit)
+
+    @pytest.mark.unit
+    def test_dof(self, frame_unit):
+        if degrees_of_freedom(frame_unit) != 0:
+            raise UnitAttributeError(
+                "The unit has {dof} degrees of freedom when 0 is required."
+                "".format(dof=degrees_of_freedom(frame_unit))
+            )
+
+    @pytest.mark.component
+    def test_initialization(self, frame_unit):
+        blk = frame_unit
+
+        # initialize
+        try:
+            blk.initialize(solver=blk._test_objs.solver, optarg=blk._test_objs.optarg)
+        except InitializationError:
+            raise InitializationError(
+                "The unit has failed to initialize successfully. Please check the output logs for more information."
+            )
+
+    @pytest.mark.component
+    def test_unit_solutions(self, frame_unit):
+        self.configure_class()
+        blk = frame_unit
+        solutions = blk._test_objs.unit_solutions
+
+        # solve unit
+        if blk._test_objs.solver is None:
+            opt = get_solver()
+        else:
+            opt = get_solver(
+                solver=blk._test_objs.solver, options=blk._test_objs.optarg
+            )
+        results = opt.solve(blk)
+
+        # check solve
+        try:
+            assert len(list(iscale.badly_scaled_var_generator(blk, zero=1e-8))) == 0
+        except AssertionError:
+            badly_scaled_var_list = iscale.badly_scaled_var_generator(blk, zero=1e-8)
+            for x in badly_scaled_var_list:
+                print(
+                    f"{x[0].name}\t{x[0].value}\tsf: {iscale.get_scaling_factor(x[0])}"
+                )
+            raise AssertionError(
+                "Check the output logs for variables that need to be re-scaled"
+            )
+        assert_optimal_termination(results)
+
+        # check results
+        for key, val in solutions.items():
+            pyo_obj = blk.find_component(key)
+            pyo_obj_value = pyo_obj()
+            try:
+                if val == 0:
+                    assert pytest.approx(pyo_obj_value, abs=1e-08) == val
+                else:
+                    assert pytest.approx(pyo_obj_value, rel=1e-03) == val
```

Comment:
You can also pass both the `abs` and `rel` argument to `pytest.approx` and it will pass as long as one of them is satisfied which is often all you need (and also handles cases where the value is very small but non-zero).